Objects/listobject.c

```diff
@@ -569,24 +566,35 @@ list_repeat(PyListObject *a, Py_ssize_t n)
     if (np == NULL)
         return NULL;
 
+    PyObject **dest = np->ob_item;
+    PyObject **dest_end = dest + size;
     if (Py_SIZE(a) == 1) {
-        items = np->ob_item;
-        elem = a->ob_item[0];
-        for (i = 0; i < n; i++) {
-            items[i] = elem;
-            Py_INCREF(elem);
+        PyObject *elem = a->ob_item[0];
+        Py_SET_REFCNT(elem, Py_REFCNT(elem) + n);
+#ifdef Py_REF_DEBUG
+        _Py_RefTotal += n;
+#endif
+        while (dest < dest_end) {
+            *dest++ = elem;
         }
     }
     else {
-        p = np->ob_item;
-        items = a->ob_item;
-        for (i = 0; i < n; i++) {
-            for (j = 0; j < Py_SIZE(a); j++) {
-                *p = items[j];
-                Py_INCREF(*p);
-                p++;
+        PyObject **src0 = a->ob_item;
+        PyObject **src_end = src0 + Py_SIZE(a);
+        PyObject **src;
+        for (src = src0; src < src_end; src++) {
+            Py_SET_REFCNT(*src, Py_REFCNT(*src) + n);
```

Comment:
Have you considered adding something like Py_INCREF_N(op, n)  that shifts the refcount by n and updates _Py_RefTotal?

I think most usages or Py_SET_REFCNT in the codebase could be replaced by that.

Reply:
Interesting! Maybe for a separate PR? For reference, here are the uses of `Py_SET_REFCNT` that Visual Studio found, including the 3 that this PR adds:

| Code                                       | File                                     | Line   | Column  | Project        | 
|--------------------------------------------|------------------------------------------|--------|---------|----------------| 
| Py_SET_REFCNT(elem, Py_REFCNT(elem) + n);  | list_repeat\\Objects\\listobject.c       | 573    | 9       | _freeze_module | 
| Py_SET_REFCNT(*src, Py_REFCNT(*src) + n);  | list_repeat\\Objects\\listobject.c       | 586    | 13      | _freeze_module | 
| Py_SET_REFCNT(def, 1);                     | list_repeat\\Objects\\moduleobject.c     | 53     | 9       | _freeze_module | 
| Py_SET_REFCNT(self, 1);                    | list_repeat\\Objects\\object.c           | 234    | 5       | _freeze_module | 
| Py_SET_REFCNT(self, Py_REFCNT(self) - 1);  | list_repeat\\Objects\\object.c           | 244    | 5       | _freeze_module | 
| Py_SET_REFCNT(self, refcnt);               | list_repeat\\Objects\\object.c           | 253    | 5       | _freeze_module | 
| Py_SET_REFCNT(op, 1);                      | list_repeat\\Objects\\object.c           | 1955   | 5       | _freeze_module | 
| Py_SET_REFCNT(*src, Py_REFCNT(*src) + n);  | list_repeat\\Objects\\tupleobject.c      | 616    | 9       | _freeze_module | 
| Py_SET_REFCNT(unicode, 3);                 | list_repeat\\Objects\\unicodeobject.c    | 1960   | 9       | _freeze_module | 
| Py_SET_REFCNT(unicode, 0);                 | list_repeat\\Objects\\unicodeobject.c    | 1966   | 9       | _freeze_module | 
| Py_SET_REFCNT(s, Py_REFCNT(s) - 2);        | list_repeat\\Objects\\unicodeobject.c    | 15623  | 5       | _freeze_module | 
| Py_SET_REFCNT(s, Py_REFCNT(s) + 1);        | list_repeat\\Objects\\unicodeobject.c    | 15684  | 13      | _freeze_module | 
| Py_SET_REFCNT(s, Py_REFCNT(s) + 2);        | list_repeat\\Objects\\unicodeobject.c    | 15692  | 13      | _freeze_module | 
| Py_SET_REFCNT(self, 1);                    | list_repeat\\Modules\\_testcapimodule.c  | 3860   | 5       | _testcapi      | 
| Py_SET_REFCNT(self, Py_REFCNT(self) - 1);  | list_repeat\\Modules\\_testcapimodule.c  | 3883   | 5       | _testcapi      | 
| Py_SET_REFCNT(self, refcnt);               | list_repeat\\Modules\\_testcapimodule.c  | 3895   | 9       | _testcapi      | 
| Py_SET_REFCNT(op, 1);                      | list_repeat\\Modules\\_testcapimodule.c  | 5003   | 5       | _testcapi      | 
| Py_SET_REFCNT(op, 1);                      | list_repeat\\Modules\\_testcapimodule.c  | 5018   | 5       | _testcapi      | 
| Py_SET_REFCNT(op, 1);                      | list_repeat\\Modules\\_testcapimodule.c  | 5038   | 5       | _testcapi      | 
| Py_SET_REFCNT(obj,  0);                    | list_repeat\\Modules\\_testcapimodule.c  | 5522   | 5       | _testcapi      |